### PR TITLE
Add `PickerAvoidingView` and `PickerStateProvider` for better iOS modal handling

### PR DESCRIPTION
**Pull request by jakub-trzebiatowski**
This PR introduces two new components, `PickerAvoidingView` and `PickerStateProvider`. These components aim to improve the user experience on iOS by ensuring that the picker is not obstructed by its own modal. The PickerAvoidingView handles proper positioning of the picker, while the PickerStateProvider manages the state and interactions between multiple pickers. This update enhances the overall usability of the picker component on iOS devices.

Note: This PR is stacked on top of https://github.com/lawnstarter/react-native-picker-select/pull/496, so that one needs to be merged first

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -214,12 +231,41 @@ export default class RNPickerSelect extends PureComponent {
...
+
+            // If TextInput is below picker modal, scroll up
+            if (textInputBottomY > modalY) {
+                // Wait until the modal animation finishes, so the scrolling is effective when PickerAvoidingView is
+                // used
+                setTimeout(() => {
+                    this.props.scrollViewRef.current.scrollTo({
+                        y: textInputBottomY - modalY + 10 + this.props.scrollViewContentOffsetY,
+                    });
+                }, IOS_MODAL_ANIMATION_DURATION_MS + 50);
```
Comment (Julesssss) [inline src/index.js:257]: Can we figure out a way of avoiding the setTimeout? Possible via a hook?

Comment (jakub-trzebiatowski) [reply src/index.js:257]: This PR was filed by mistake. It was meant for the upstream. I fixed that. Upstream PR: https://github.com/lawnstarter/react-native-picker-select/pull/507

Comment (jakub-trzebiatowski) [reply src/index.js:257]: Let's continue this thread [here](https://github.com/Expensify/react-native-picker-select/pull/8/files#r1166618576)

**Comment by jakub-trzebiatowski** [comment]
PR filed by mistake, was meant for the upstream

**Pull request closed without merging**